examples/src/test/scala/JedisExample.scala

```diff
@@ -58,83 +60,108 @@ object DataSources {
-      Sync[F].pure({
-        in.writeObject(v)
-        in.flush()
-        baos.toByteArray
+  def serialize[F[_], A](obj: A)(
+      implicit S: Sync[F]
+  ): F[ByteArray] = {
+    byteOutputStream
+      .mproduct(outputStream(_))
+      .use({
+        case (byte, out) =>
+          S.delay {
+            out.writeObject(obj)
+            out.flush()
+            byte.toByteArray
+          }
       })
-    })({ in =>
-      Sync[F].delay(in.close())
-    })
   }
 
-  def deserialize[F[_]: Sync, I, A](bs: Option[Array[Byte]]): F[Option[A]] =
-    bs match {
-      case None => Sync[F].pure(None)
-      case Some(raw) => {
-        val bais = new ByteArrayInputStream(raw)
-        val oos  = new ObjectInputStream(bais)
-        Sync[F].bracket(
-          Sync[F].pure((bais, oos))
-        )({
-          case (b, o) =>
-            Sync[F].pure({
-              val res = o.readObject()
-              Option(res.asInstanceOf[A])
-            })
-        })({
-          case (b, o) =>
-            Sync[F].delay({
-              b.close()
-              o.close()
-            })
-        })
-      }
-    }
-
+  def deserialize[F[_], A](bin: ByteArray)(
+      implicit S: Sync[F]
+  ): F[Option[A]] = {
+    byteInputStream(bin)
+      .mproduct(inputStream(_))
+      .use({
+        case (byte, in) =>
+          S.delay {
+            val obj = in.readObject()
+            Try(obj.asInstanceOf[A]).toOption
+          }
+      })
+  }
 }
 
 case class RedisCache[F[_]: Sync](host: String) extends DataCache[F] {
-  private val binaryClient = new BinaryJedis(host)
+  private val pool = new JedisPool(host)
 
-  def client: F[BinaryJedis] =
-    Sync[F].pure(binaryClient)
+  def connection: Resource[F, Jedis] =
+    Resource.fromAutoCloseable(Sync[F].delay(pool.getResource))
 
   private def get(i: Array[Byte]): F[Option[Array[Byte]]] =
-    Sync[F].bracket(client)({ c =>
-      Sync[F].pure(Option(c.get(i)))
-    })({ in =>
-      Sync[F].pure(in.close())
-    })
+    connection.use(c => Sync[F].delay(Option(c.get(i))))
 
   private def set(i: Array[Byte], v: Array[Byte]): F[Unit] =
-    Sync[F].bracket(client)({ c =>
-      Sync[F].pure(c.set(i, v)).void
-    })({ in =>
-      Sync[F].pure(in.close())
-    })
+    connection.use(c => Sync[F].delay(c.set(i, v)).void)
 
-  private def identity[I, A](i: I, data: Data[I, A]): Array[Byte] =
+  private def bulkSet(ivs: List[(Array[Byte], Array[Byte])]): F[Unit] =
+    connection.use(c =>
+      Sync[F].delay({
+        val pipe = c.pipelined
+        ivs.foreach(i => pipe.set(i._1, i._2))
+        pipe.sync
+      }))
+
+  private def cacheId[I, A](i: I, data: Data[I, A]): Array[Byte] =
     Binary.fromString(s"${data.identity} ${i}")
 
-  def lookup[I, A](i: I, data: Data[I, A]): F[Option[A]] =
-    get(identity(i, data)) >>= { case (raw) => Binary.deserialize(raw) }
+  override def lookup[I, A](i: I, data: Data[I, A]): F[Option[A]] =
+    get(cacheId(i, data)).flatMap(_ match {
```

Comment:
You do not need the `_ match`. Just the two cases between braces make up the literal of a `PartialFunction`, which (regrettably) is also a `Function`